### PR TITLE
[BEAM-545] Promote JobName to PipelineOptions

### PR DESCRIPTION
**Pull request by peihe**
This RP also fixed https://issues.apache.org/jira/browse/BEAM-483 "Generated job names easily collide"

### CODE REVIEW

**Comment by peihe** [comment]
R: @dhalperi

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/ApplicationNameOptions.java [hunk 1]
```diff
@@ -31,4 +40,39 @@
...
+   * [a-z]([-a-z0-9]*[a-z0-9])?.
+   */
+  public static class NormalizedUniqueNameFactory implements DefaultValueFactory<String> {
+    private static final DateTimeFormatter FORMATTER =
+        DateTimeFormat.forPattern("MMddHHmmss").withZone(DateTimeZone.UTC);
+
+    @Override
+    public String create(PipelineOptions options) {
+      String appName = options.as(ApplicationNameOptions.class).getAppName();
+      String normalizedAppName = appName == null || appName.length() == 0 ? "BeamApp"
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/ApplicationNameOptions.java]: `Strings.isNullOrEmpty`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/ApplicationNameOptions.java [hunk 2]
```diff
@@ -31,4 +40,39 @@
       + "constructs the PipelineOptions via the PipelineOptionsFactory.")
   String getAppName();
   void setAppName(String value);
+
+  @Description("A normalized unique name that is used to name anything related to the pipeline."
+      + "It defaults to ApplicationName-UserName-Date-RandomInteger")
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/ApplicationNameOptions.java]: DateTime?

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java [hunk 3]
```diff
@@ -98,21 +94,9 @@
 
 
   class JobNameFactory implements DefaultValueFactory<String> {
```
Comment (dhalperi) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java]: used for `JobName` above, but is that actually still needed? or can they use `getNormalizedUniqueName()` instead?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java [hunk 4]
```diff
@@ -113,21 +106,9 @@
...
-      String normalizedAppName = appName == null || appName.length() == 0 ? "dataflow"
-          : appName.toLowerCase()
-                   .replaceAll("[^a-z0-9]", "0")
-                   .replaceAll("^[^a-z]", "a");
-      String userName = MoreObjects.firstNonNull(System.getProperty("user.name"), "");
-      String normalizedUserName = userName.toLowerCase()
-                                          .replaceAll("[^a-z0-9]", "0");
-      String datePart = FORMATTER.print(DateTimeUtils.currentTimeMillis());
-      return normalizedAppName + "-" + normalizedUserName + "-" + datePart;
+      return options.as(ApplicationNameOptions.class).getNormalizedUniqueName();
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: Same question about Job name.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/ApplicationNameOptions.java [hunk 5]
```diff
@@ -31,4 +40,39 @@
...
+  @Default.InstanceFactory(NormalizedUniqueNameFactory.class)
+  String getNormalizedUniqueName();
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/ApplicationNameOptions.java]: Why isn't this just `getJobName`?

**Comment by dhalperi** [comment]
Status?

**Comment by peihe** [comment]
PR updated based on the proposal sent to dev@beam.incubator.apache.org mailing list

PTAL

**Pull request closed without merging**